File: airs_experiment/R/run_03_invariance.R
#!/usr/bin/env Rscript
# =============================================================================
# AIRS Phase 3: Measurement Invariance Testing — R port
# Mirrors: python/run_03_invariance.py
#
# Academic vs Professional groups: configural, metric, scalar invariance.
# Method: lavaan multi-group CFA with semTools measEq.syntax
# Outputs: data/invariance_results.json, tables/invariance_*.csv
# =============================================================================

library(lavaan)
library(semTools)
library(psych)
library(jsonlite)

cat(paste(rep("=", 90), collapse = ""), "\n")
cat("PHASE 3: MEASUREMENT INVARIANCE TESTING\n")
cat(paste(rep("=", 90), collapse = ""), "\n")

dir.create("tables", showWarnings = FALSE)

# =============================================================================
# LOAD DATA
# =============================================================================

df <- read.csv("data/AIRS_clean.csv", stringsAsFactors = FALSE)
cfa_summary <- fromJSON("tables/cfa_summary.json")

cat(sprintf("\nFull sample N = %d\n", nrow(df)))

# Create population grouping
df$Population <- ifelse(df$Role_Category %in% c("FT_Student", "PT_Student"),
    "Academic", "Professional"
)
cat("Group sizes:\n")
print(table(df$Population))

# =============================================================================
# CFA MODEL SYNTAX (8-factor, 16 items)
# =============================================================================

cfa_model <- "
  PerfExp   =~ PE1 + PE2
  EffortExp =~ EE1 + EE2
  SocialInf =~ SI1 + SI2
  FacilCond =~ FC1 + FC2
  HedonicMot =~ HM1 + HM2
  PriceValue =~ PV1 + PV2
  Habit     =~ HB1 + HB2
  AITrust   =~ TR1 + TR2
"

# =============================================================================
# SEPARATE GROUP CFAs
# =============================================================================

cat("\n", paste(rep("=", 90), collapse = ""), "\n")
cat("SEPARATE GROUP CFA MODELS\n")
cat(paste(rep("=", 90), collapse = ""), "\n")

for (grp in c("Academic", "Professional")) {
    df_grp <- df[df$Population == grp, ]
    cat(sprintf("\n--- %s (n=%d) ---\n", grp, nrow(df_grp)))

    fit_grp <- tryCatch(
        {
            cfa(cfa_model, data = df_grp, estimator = "ML")
        },
        error = function(e) {
            cat("  ERROR:", e$message, "\n")
            NULL
        }
    )

    if (!is.null(fit_grp) && lavInspect(fit_grp, "converged")) {
        fm <- fitMeasures(fit_grp, c("chisq", "df", "cfi", "tli", "rmsea", "srmr"))
        cat(sprintf(
            "  CFI = %.4f, TLI = %.4f, RMSEA = %.4f, SRMR = %.4f\n",
            fm["cfi"], fm["tli"], fm["rmsea"], fm["srmr"]
        ))

        # Reliability per group
        factor_items <- list(
            PerfExp = c("PE1", "PE2"), EffortExp = c("EE1", "EE2"),
            SocialInf = c("SI1", "SI2"), FacilCond = c("FC1", "FC2"),
            HedonicMot = c("HM1", "HM2"), PriceValue = c("PV1", "PV2"),
            Habit = c("HB1", "HB2"), AITrust = c("TR1", "TR2")
        )
        cat("  Reliability (alpha):\n")
        for (fname in names(factor_items)) {
            itms <- factor_items[[fname]]
            alpha_val <- psych::alpha(df_grp[itms], check.keys = FALSE)$total$raw_alpha
            cat(sprintf("    %s: %.3f\n", fname, alpha_val))
        }
    } else {
        cat("  Model did not converge.\n")
    }
}

# =============================================================================
# MULTI-GROUP INVARIANCE TESTING
# =============================================================================

cat("\n", paste(rep("=", 90), collapse = ""), "\n")
cat("MULTI-GROUP INVARIANCE TESTING\n")
cat(paste(rep("=", 90), collapse = ""), "\n")

# Configural invariance (same structure, all parameters free)
fit_configural <- cfa(cfa_model, data = df, group = "Population", estimator = "ML")
fm_config <- fitMeasures(fit_configural, c("chisq", "df", "cfi", "tli", "rmsea", "srmr", "aic", "bic"))

cat("\n--- Configural Invariance ---\n")
cat(sprintf(
    "  CFI = %.4f, RMSEA = %.4f, SRMR = %.4f\n",
    fm_config["cfi"], fm_config["rmsea"], fm_config["srmr"]
))

# Metric invariance (loadings constrained equal)
fit_metric <- cfa(cfa_model,
    data = df, group = "Population",
    group.equal = "loadings", estimator = "ML"
)
fm_metric <- fitMeasures(fit_metric, c("chisq", "df", "cfi", "tli", "rmsea", "srmr", "aic", "bic"))

cat("\n--- Metric Invariance ---\n")
cat(sprintf(
    "  CFI = %.4f, RMSEA = %.4f, SRMR = %.4f\n",
    fm_metric["cfi"], fm_metric["rmsea"], fm_metric["srmr"]
))

# Scalar invariance (loadings + intercepts constrained)
fit_scalar <- cfa(cfa_model,
    data = df, group = "Population",
    group.equal = c("loadings", "intercepts"), estimator = "ML"
)
fm_scalar <- fitMeasures(fit_scalar, c("chisq", "df", "cfi", "tli", "rmsea", "srmr", "aic", "bic"))

cat("\n--- Scalar Invariance ---\n")
cat(sprintf(
    "  CFI = %.4f, RMSEA = %.4f, SRMR = %.4f\n",
    fm_scalar["cfi"], fm_scalar["rmsea"], fm_scalar["srmr"]
))

# =============================================================================
# CHI-SQUARE DIFFERENCE TESTS
# =============================================================================

cat("\n", paste(rep("=", 90), collapse = ""), "\n")
cat("MODEL COMPARISON (Chi-Square Difference Tests)\n")
cat(paste(rep("=", 90), collapse = ""), "\n")

# Configural vs Metric
diff_cm <- anova(fit_configural, fit_metric)
cat("\nConfigural vs Metric:\n")
print(diff_cm)
delta_cfi_cm <- fm_config["cfi"] - fm_metric["cfi"]
cat(sprintf(
    "  Delta_CFI = %.4f %s\n", delta_cfi_cm,
    ifelse(abs(delta_cfi_cm) <= 0.01, "[INVARIANCE SUPPORTED]", "[INVARIANCE NOT SUPPORTED]")
))

# Metric vs Scalar
diff_ms <- anova(fit_metric, fit_scalar)
cat("\nMetric vs Scalar:\n")
print(diff_ms)
delta_cfi_ms <- fm_metric["cfi"] - fm_scalar["cfi"]
cat(sprintf(
    "  Delta_CFI = %.4f %s\n", delta_cfi_ms,
    ifelse(abs(delta_cfi_ms) <= 0.01, "[INVARIANCE SUPPORTED]", "[INVARIANCE NOT SUPPORTED]")
))

# =============================================================================
# LOADING COMPARISON ACROSS GROUPS
# =============================================================================

cat("\n", paste(rep("=", 90), collapse = ""), "\n")
cat("LOADING COMPARISON (Configural Model)\n")
cat(paste(rep("=", 90), collapse = ""), "\n")

# Extract per-group standardized loadings from configural model
std_config <- standardizedSolution(fit_configural)
loadings_config <- std_config[
    std_config$op == "=~",
    c("lhs", "rhs", "est.std", "group")
]

# Group labels
grp_labels <- lavInspect(fit_configural, "group.label")

loading_comparison <- data.frame(
    Factor = character(), Item = character(),
    Loading_G1 = numeric(), Loading_G2 = numeric(),
    Diff = numeric(), stringsAsFactors = FALSE
)

for (fac in unique(loadings_config$lhs)) {
    for (itm in unique(loadings_config$rhs[loadings_config$lhs == fac])) {
        l1 <- loadings_config$est.std[loadings_config$lhs == fac &
            loadings_config$rhs == itm &
            loadings_config$group == 1]
        l2 <- loadings_config$est.std[loadings_config$lhs == fac &
            loadings_config$rhs == itm &
            loadings_config$group == 2]
        if (length(l1) > 0 && length(l2) > 0) {
            loading_comparison <- rbind(loading_comparison, data.frame(
                Factor = fac, Item = itm,
                Loading_G1 = l1, Loading_G2 = l2,
                Diff = abs(l1 - l2), stringsAsFactors = FALSE
            ))
        }
    }
}

colnames(loading_comparison)[3:4] <- paste0("Loading_", grp_labels)
cat("\n")
num_lc <- sapply(loading_comparison, is.numeric)
print(cbind(loading_comparison[, !num_lc, drop = FALSE], round(loading_comparison[, num_lc, drop = FALSE], 4)))

write.csv(loading_comparison, "tables/invariance_loading_comparison.csv", row.names = FALSE)
cat("\nExported: tables/invariance_loading_comparison.csv\n")

# =============================================================================
# GROUP MEAN COMPARISONS (Cohen's d)
# =============================================================================

cat("\n", paste(rep("=", 90), collapse = ""), "\n")
cat("GROUP MEAN COMPARISONS (Cohen's d)\n")
cat(paste(rep("=", 90), collapse = ""), "\n")

factor_items <- list(
    PerfExp = c("PE1", "PE2"), EffortExp = c("EE1", "EE2"),
    SocialInf = c("SI1", "SI2"), FacilCond = c("FC1", "FC2"),
    HedonicMot = c("HM1", "HM2"), PriceValue = c("PV1", "PV2"),
    Habit = c("HB1", "HB2"), AITrust = c("TR1", "TR2")
)

df_acad <- df[df$Population == "Academic", ]
df_prof <- df[df$Population == "Professional", ]

mean_comparison <- data.frame(
    Factor = character(), Mean_Academic = numeric(), SD_Academic = numeric(),
    Mean_Professional = numeric(), SD_Professional = numeric(),
    Cohens_d = numeric(), t_value = numeric(), p_value = numeric(),
    stringsAsFactors = FALSE
)

for (fname in names(factor_items)) {
    items <- factor_items[[fname]]
    score_acad <- rowMeans(df_acad[, items])
    score_prof <- rowMeans(df_prof[, items])

    t_result <- t.test(score_acad, score_prof)
    # Cohen's d
    pooled_sd <- sqrt(((length(score_acad) - 1) * sd(score_acad)^2 +
        (length(score_prof) - 1) * sd(score_prof)^2) /
        (length(score_acad) + length(score_prof) - 2))
    d <- (mean(score_acad) - mean(score_prof)) / pooled_sd

    mean_comparison <- rbind(mean_comparison, data.frame(
        Factor = fname,
        Mean_Academic = mean(score_acad), SD_Academic = sd(score_acad),
        Mean_Professional = mean(score_prof), SD_Professional = sd(score_prof),
        Cohens_d = d,
        t_value = t_result$statistic,
        p_value = t_result$p.value,
        stringsAsFactors = FALSE
    ))

    effect <- ifelse(abs(d) >= 0.80, "LARGE",
        ifelse(abs(d) >= 0.50, "MEDIUM",
            ifelse(abs(d) >= 0.20, "SMALL", "NEGLIGIBLE")
        )
    )
    sig <- ifelse(t_result$p.value < 0.05, "*", "")
    cat(sprintf(
        "  %s: Acad=%.2f, Prof=%.2f, d=%.3f [%s] %s\n",
        fname, mean(score_acad), mean(score_prof), d, effect, sig
    ))
}

write.csv(mean_comparison, "tables/invariance_mean_comparison.csv", row.names = FALSE)
cat("\nExported: tables/invariance_mean_comparison.csv\n")

# =============================================================================
# EXPORT RESULTS JSON
# =============================================================================

invariance_results <- list(
    analysis_date = format(Sys.time(), "%Y-%m-%dT%H:%M:%S"),
    n_total = nrow(df),
    n_academic = nrow(df_acad),
    n_professional = nrow(df_prof),
    configural = as.list(round(fm_config, 6)),
    metric = as.list(round(fm_metric, 6)),
    scalar = as.list(round(fm_scalar, 6)),
    delta_cfi_configural_metric = round(as.numeric(delta_cfi_cm), 4),
    delta_cfi_metric_scalar = round(as.numeric(delta_cfi_ms), 4),
    metric_invariance_supported = abs(as.numeric(delta_cfi_cm)) <= 0.01,
    scalar_invariance_supported = abs(as.numeric(delta_cfi_ms)) <= 0.01,
    mean_comparisons = lapply(seq_len(nrow(mean_comparison)), function(i) {
        as.list(mean_comparison[i, ])
    })
)

write_json(invariance_results, "data/invariance_results.json",
    pretty = TRUE, auto_unbox = TRUE
)
cat("\nExported: data/invariance_results.json\n")

# =============================================================================
# INVARIANCE SUMMARY TABLE
# =============================================================================

invariance_table <- data.frame(
    Model = c("Configural", "Metric", "Scalar"),
    chi2 = c(fm_config["chisq"], fm_metric["chisq"], fm_scalar["chisq"]),
    df = c(fm_config["df"], fm_metric["df"], fm_scalar["df"]),
    CFI = c(fm_config["cfi"], fm_metric["cfi"], fm_scalar["cfi"]),
    RMSEA = c(fm_config["rmsea"], fm_metric["rmsea"], fm_scalar["rmsea"]),
    SRMR = c(fm_config["srmr"], fm_metric["srmr"], fm_scalar["srmr"]),
    stringsAsFactors = FALSE
)
invariance_table$Delta_CFI <- c(
    NA,
    as.numeric(delta_cfi_cm),
    as.numeric(delta_cfi_ms)
)
invariance_table$Supported <- c(
    TRUE,
    abs(as.numeric(delta_cfi_cm)) <= 0.01,
    abs(as.numeric(delta_cfi_ms)) <= 0.01
)

write.csv(invariance_table, "tables/invariance_summary.csv", row.names = FALSE)

cat("\n", paste(rep("=", 90), collapse = ""), "\n")
cat("INVARIANCE SUMMARY\n")
cat(paste(rep("=", 90), collapse = ""), "\n")
num_it <- sapply(invariance_table[, -8], is.numeric)
print(cbind(invariance_table[, -8][, !num_it, drop = FALSE], round(invariance_table[, -8][, num_it, drop = FALSE], 4)))

cat("\n", paste(rep("=", 90), collapse = ""), "\n")
cat("PHASE 3 COMPLETE\n")
cat(paste(rep("=", 90), collapse = ""), "\n")

File: airs_experiment/python/run_03_invariance.py
#!/usr/bin/env python3
"""
AIRS Phase 3: Measurement Invariance Testing
Refactored from 03_Measurement_Invariance.ipynb

Tests measurement equivalence across Academic vs Professional populations.
Outputs: data/invariance_results.json, tables/invariance_*.csv,
         plots/measurement_invariance_*.png
"""

import json
import os
import warnings
from pathlib import Path

import numpy as np
import pandas as pd
import matplotlib
matplotlib.use('Agg')
import matplotlib.pyplot as plt
import seaborn as sns
from scipy import stats

import semopy
from semopy import Model
import pingouin as pg

os.environ['KMP_DUPLICATE_LIB_OK'] = 'True'
warnings.filterwarnings('ignore')

RANDOM_SEED = 67


def build_cfa_syntax(structure):
    """Build semopy CFA syntax from structure dict."""
    lines = []
    factors = list(structure.keys())
    for factor, items in structure.items():
        lines.append(f"{factor} =~ {' + '.join(items)}")
    for i, f1 in enumerate(factors):
        for f2 in factors[i+1:]:
            lines.append(f"{f1} ~~ {f2}")
    return '\n'.join(lines)


def fit_cfa_model(df, syntax, group_name):
    """Fit CFA and return fit statistics."""
    model = Model(syntax)
    model.fit(df)
    fit_stats = semopy.calc_stats(model)

    chi2 = float(fit_stats.loc['Value', 'chi2']) if 'chi2' in fit_stats.columns else np.nan
    df_val = float(fit_stats.loc['Value', 'DoF']) if 'DoF' in fit_stats.columns else np.nan
    cfi = float(fit_stats.loc['Value', 'CFI']) if 'CFI' in fit_stats.columns else np.nan
    tli = float(fit_stats.loc['Value', 'TLI']) if 'TLI' in fit_stats.columns else np.nan
    rmsea = float(fit_stats.loc['Value', 'RMSEA']) if 'RMSEA' in fit_stats.columns else np.nan
    aic = float(fit_stats.loc['Value', 'AIC']) if 'AIC' in fit_stats.columns else np.nan
    bic = float(fit_stats.loc['Value', 'BIC']) if 'BIC' in fit_stats.columns else np.nan

    print(f"\n{group_name} CFA (N={len(df)}): chi2({df_val:.0f})={chi2:.2f}, "
          f"CFI={cfi:.3f}, TLI={tli:.3f}, RMSEA={rmsea:.3f}")

    return {'model': model, 'chi2': chi2, 'df': df_val, 'cfi': cfi, 'tli': tli,
            'rmsea': rmsea, 'aic': aic, 'bic': bic, 'n': len(df)}


def extract_loadings(model, structure):
    """Extract standardized factor loadings."""
    params = model.inspect(std_est=True)
    loadings = params[params['op'] == '~'].copy()
    loading_dict = {}
    for _, row in loadings.iterrows():
        factor = row['rval']
        item = row['lval']
        est = row['Est. Std']
        loading_dict[(factor, item)] = float(est) if pd.notna(est) else np.nan
    return loading_dict


def main():
    np.random.seed(RANDOM_SEED)

    print("=" * 90)
    print("PHASE 3: MEASUREMENT INVARIANCE TESTING")
    print("=" * 90)

    Path('plots').mkdir(exist_ok=True)
    Path('tables').mkdir(exist_ok=True)
    sns.set_style('whitegrid')

    # ── Load Data ─────────────────────────────────────────────────────────
    df_efa = pd.read_csv('data/AIRS_experiment.csv')
    df_cfa = pd.read_csv('data/AIRS_holdout.csv')
    df_full = pd.concat([df_efa, df_cfa], ignore_index=True)
    print(f"\nFull sample: N = {len(df_full)}")

    # ── Load Model Structure ──────────────────────────────────────────────
    try:
        with open('tables/cfa_summary.json', 'r') as f:
            cfa_results = json.load(f)
        MODEL_D_STRUCTURE = cfa_results['factor_structure']
        cfa_reliability = {item['Factor']: item['Alpha'] for item in cfa_results.get('reliability', [])}
        print(f"Loaded CFA structure: {len(MODEL_D_STRUCTURE)} factors")
    except FileNotFoundError:
        MODEL_D_STRUCTURE = {
            'PerfExp': ['PE1', 'PE2'], 'EffortExp': ['EE1', 'EE2'],
            'SocialInf': ['SI1', 'SI2'], 'FacilCond': ['FC1', 'FC2'],
            'HedonicMot': ['HM1', 'HM2'], 'PriceValue': ['PV1', 'PV2'],
            'Habit': ['HB1', 'HB2'], 'AITrust': ['TR1', 'TR2'],
        }
        cfa_reliability = {}
        print("Using hardcoded fallback structure (8 factors)")

    MODEL_D_ITEMS = [item for items in MODEL_D_STRUCTURE.values() for item in items]

    for factor, items in MODEL_D_STRUCTURE.items():
        alpha_str = f" (a={cfa_reliability[factor]:.2f})" if factor in cfa_reliability else ""
        print(f"  {factor}: {items}{alpha_str}")

    # ── Create Groups ─────────────────────────────────────────────────────
    academic_roles = ['Full time student', 'Part time student']
    df_full['Population'] = df_full['Role'].apply(
        lambda x: 'Academic' if x in academic_roles else 'Professional'
    )
    df_academic = df_full[df_full['Population'] == 'Academic'].copy()
    df_professional = df_full[df_full['Population'] == 'Professional'].copy()

    print(f"\nAcademic: N = {len(df_academic)} ({len(df_academic)/len(df_full)*100:.1f}%)")
    print(f"Professional: N = {len(df_professional)} ({len(df_professional)/len(df_full)*100:.1f}%)")

    # ── Baseline CFAs ─────────────────────────────────────────────────────
    MODEL_SYNTAX = build_cfa_syntax(MODEL_D_STRUCTURE)
    print(f"\n{'=' * 90}")
    print("BASELINE CFA (Separate Groups)")
    print(f"{'=' * 90}")

    academic_fit = fit_cfa_model(df_academic[MODEL_D_ITEMS], MODEL_SYNTAX, "Academic")
    professional_fit = fit_cfa_model(df_professional[MODEL_D_ITEMS], MODEL_SYNTAX, "Professional")

    # ── Configural Invariance ─────────────────────────────────────────────
    configural_pass = (
        academic_fit['cfi'] >= 0.90 and professional_fit['cfi'] >= 0.90 and
        academic_fit['rmsea'] <= 0.08 and professional_fit['rmsea'] <= 0.08
    )
    print(f"\nConfigural Invariance: {'SUPPORTED' if configural_pass else 'QUESTIONABLE'}")

    # ── Extract Loadings ──────────────────────────────────────────────────
    academic_loadings = extract_loadings(academic_fit['model'], MODEL_D_STRUCTURE)
    professional_loadings = extract_loadings(professional_fit['model'], MODEL_D_STRUCTURE)

    loading_comparison = []
    for factor, items in MODEL_D_STRUCTURE.items():
        for item in items:
            acad_load = academic_loadings.get((factor, item), np.nan)
            prof_load = professional_loadings.get((factor, item), np.nan)
            diff = abs(acad_load - prof_load) if not np.isnan(acad_load) and not np.isnan(prof_load) else np.nan
            loading_comparison.append({
                'Factor': factor, 'Item': item,
                'Academic_Lambda': acad_load, 'Professional_Lambda': prof_load,
                'Difference': diff
            })
    loading_df = pd.DataFrame(loading_comparison)

    print(f"\n{'=' * 90}")
    print("FACTOR LOADINGS BY GROUP")
    print(f"{'=' * 90}")
    print(loading_df.to_string(index=False, float_format=lambda x: f"{x:.3f}"))

    # ── Metric Invariance ─────────────────────────────────────────────────
    mean_diff = loading_df['Difference'].mean()
    max_diff = loading_df['Difference'].max()
    items_over_threshold = int((loading_df['Difference'] > 0.10).sum())
    metric_pass = max_diff < 0.15 and items_over_threshold <= 2

    print(f"\n{'=' * 90}")
    print("METRIC INVARIANCE")
    print(f"{'=' * 90}")
    print(f"Mean |Delta-lambda| = {mean_diff:.3f}")
    print(f"Max  |Delta-lambda| = {max_diff:.3f}")
    print(f"Items > 0.10 threshold: {items_over_threshold}/{len(loading_df)}")
    print(f"{'SUPPORTED' if metric_pass else 'PARTIAL INVARIANCE'}")

    # ── Loading Plots ─────────────────────────────────────────────────────
    fig, axes = plt.subplots(1, 2, figsize=(14, 6))

    ax1 = axes[0]
    x = np.arange(len(loading_df))
    width = 0.35
    ax1.bar(x - width/2, loading_df['Academic_Lambda'], width, label='Academic', color='steelblue', alpha=0.8)
    ax1.bar(x + width/2, loading_df['Professional_Lambda'], width, label='Professional', color='coral', alpha=0.8)
    ax1.set_xticks(x)
    ax1.set_xticklabels(loading_df['Item'], rotation=45, ha='right')
    ax1.set_ylabel('Standardized Loading')
    ax1.set_title('Factor Loadings by Population')
    ax1.legend()
    ax1.axhline(y=0.70, color='green', linestyle='--', alpha=0.5)
    ax1.axhline(y=0.50, color='orange', linestyle='--', alpha=0.5)
    ax1.set_ylim(0, 1.0)

    ax2 = axes[1]
    colors = ['red' if d > 0.10 else 'green' for d in loading_df['Difference']]
    ax2.scatter(loading_df['Academic_Lambda'], loading_df['Professional_Lambda'], c=colors, s=100, alpha=0.7)
    ax2.plot([0, 1], [0, 1], 'k--', alpha=0.5)
    for i, row in loading_df.iterrows():
        ax2.annotate(row['Item'], (row['Academic_Lambda'], row['Professional_Lambda']),
                     fontsize=8, alpha=0.7, xytext=(3, 3), textcoords='offset points')
    ax2.set_xlabel('Academic Loading')
    ax2.set_ylabel('Professional Loading')
    ax2.set_title('Loading Agreement')
    ax2.set_xlim(0.4, 1.0)
    ax2.set_ylim(0.4, 1.0)

    plt.tight_layout()
    plt.savefig('plots/measurement_invariance_loadings.png', dpi=150, bbox_inches='tight')
    plt.close()
    print("\nSaved: plots/measurement_invariance_loadings.png")

    # ── Reliability by Group ──────────────────────────────────────────────
    def calc_reliability(df, structure):
        results = []
        for factor, items in structure.items():
            alpha = pg.cronbach_alpha(df[items])[0]
            results.append({'Factor': factor, 'Alpha': alpha})
        return pd.DataFrame(results)

    acad_rel = calc_reliability(df_academic, MODEL_D_STRUCTURE)
    prof_rel = calc_reliability(df_professional, MODEL_D_STRUCTURE)
    reliability_comparison = acad_rel.merge(prof_rel, on='Factor', suffixes=('_Academic', '_Professional'))
    reliability_comparison['Difference'] = abs(
        reliability_comparison['Alpha_Academic'] - reliability_comparison['Alpha_Professional']
    )

    print(f"\n{'=' * 90}")
    print("RELIABILITY BY GROUP")
    print(f"{'=' * 90}")
    print(reliability_comparison.to_string(index=False, float_format=lambda x: f"{x:.3f}"))

    # ── Mean Comparisons ──────────────────────────────────────────────────
    mean_comparison = []
    for factor, items in MODEL_D_STRUCTURE.items():
        acad_scores = df_academic[items].mean(axis=1)
        prof_scores = df_professional[items].mean(axis=1)
        t_stat, p_value = stats.ttest_ind(acad_scores, prof_scores)
        pooled_std = np.sqrt(((len(acad_scores)-1)*acad_scores.std()**2 +
                              (len(prof_scores)-1)*prof_scores.std()**2) /
                             (len(acad_scores) + len(prof_scores) - 2))
        cohens_d = (acad_scores.mean() - prof_scores.mean()) / pooled_std if pooled_std > 0 else 0

        if abs(cohens_d) < 0.20: effect = "negligible"
        elif abs(cohens_d) < 0.50: effect = "small"
        elif abs(cohens_d) < 0.80: effect = "medium"
        else: effect = "large"

        mean_comparison.append({
            'Factor': factor,
            'M_Academic': acad_scores.mean(), 'SD_Academic': acad_scores.std(),
            'M_Professional': prof_scores.mean(), 'SD_Professional': prof_scores.std(),
            't': t_stat, 'p': p_value, "Cohen's d": cohens_d, 'Effect': effect,
            'Sig': '***' if p_value < 0.001 else '**' if p_value < 0.01 else '*' if p_value < 0.05 else ''
        })
    mean_df = pd.DataFrame(mean_comparison)

    n_sig_means = int((mean_df['p'] < 0.05).sum())
    n_medium_large = int((mean_df["Cohen's d"].abs() >= 0.50).sum())

    print(f"\n{'=' * 90}")
    print("MEAN COMPARISONS: ACADEMIC VS PROFESSIONAL")
    print(f"{'=' * 90}")
    print(mean_df[['Factor', 'M_Academic', 'SD_Academic', 'M_Professional', 'SD_Professional',
                    't', 'p', "Cohen's d", 'Effect', 'Sig']]
          .to_string(index=False, float_format=lambda x: f"{x:.3f}" if isinstance(x, float) else x))
    print(f"\nSignificant diffs: {n_sig_means}/{len(mean_df)}, Medium/Large effects: {n_medium_large}/{len(mean_df)}")

    # ── Mean Difference Plots ─────────────────────────────────────────────
    fig, axes = plt.subplots(1, 2, figsize=(14, 6))

    ax1 = axes[0]
    x = np.arange(len(mean_df))
    width = 0.35
    ax1.bar(x - width/2, mean_df['M_Academic'], width,
            yerr=mean_df['SD_Academic']/np.sqrt(len(df_academic)),
            label='Academic', color='steelblue', alpha=0.8, capsize=3)
    ax1.bar(x + width/2, mean_df['M_Professional'], width,
            yerr=mean_df['SD_Professional']/np.sqrt(len(df_professional)),
            label='Professional', color='coral', alpha=0.8, capsize=3)
    for i, row in mean_df.iterrows():
        if row['Sig']:
            max_y = max(row['M_Academic'], row['M_Professional']) + 0.3
            ax1.text(i, max_y, row['Sig'], ha='center', fontsize=12, fontweight='bold')
    ax1.set_xticks(x)
    ax1.set_xticklabels(mean_df['Factor'], rotation=45, ha='right')
    ax1.set_ylabel('Mean Score (1-5)')
    ax1.set_title('Construct Means by Population')
    ax1.legend()
    ax1.set_ylim(1, 5.5)

    ax2 = axes[1]
    colors = ['steelblue' if d > 0 else 'coral' for d in mean_df["Cohen's d"]]
    ax2.barh(mean_df['Factor'], mean_df["Cohen's d"], color=colors, alpha=0.8)
    ax2.axvline(x=0.20, color='gray', linestyle='--', alpha=0.5)
    ax2.axvline(x=-0.20, color='gray', linestyle='--', alpha=0.5)
    ax2.axvline(x=0, color='black', linewidth=1)
    ax2.set_xlabel("Cohen's d")
    ax2.set_title('Effect Sizes for Group Differences')
    ax2.set_xlim(-1.0, 1.0)

    plt.tight_layout()
    plt.savefig('plots/measurement_invariance_means.png', dpi=150, bbox_inches='tight')
    plt.close()
    print("Saved: plots/measurement_invariance_means.png")

    # ── Export ────────────────────────────────────────────────────────────
    loading_df.to_csv('tables/invariance_loadings.csv', index=False)
    reliability_comparison.to_csv('tables/invariance_reliability.csv', index=False)
    mean_df.to_csv('tables/invariance_means.csv', index=False)

    invariance_summary = {
        'analysis_date': pd.Timestamp.now().isoformat(),
        'n_academic': len(df_academic),
        'n_professional': len(df_professional),
        'academic_fit': {
            'chi2': float(academic_fit['chi2']), 'df': float(academic_fit['df']),
            'cfi': float(academic_fit['cfi']), 'tli': float(academic_fit['tli']),
            'rmsea': float(academic_fit['rmsea'])
        },
        'professional_fit': {
            'chi2': float(professional_fit['chi2']), 'df': float(professional_fit['df']),
            'cfi': float(professional_fit['cfi']), 'tli': float(professional_fit['tli']),
            'rmsea': float(professional_fit['rmsea'])
        },
        'configural_invariance': bool(configural_pass),
        'metric_invariance': bool(metric_pass),
        'mean_loading_diff': float(mean_diff),
        'max_loading_diff': float(max_diff),
        'n_sig_mean_diffs': n_sig_means,
        'n_medium_large_effects': n_medium_large
    }
    with open('data/invariance_results.json', 'w') as f:
        json.dump(invariance_summary, f, indent=2, default=str)

    print("\nExported: tables/invariance_*.csv, data/invariance_results.json")

    print(f"\n{'=' * 90}")
    print("PHASE 3 COMPLETE")
    print(f"{'=' * 90}")


if __name__ == '__main__':
    main()
